pkg/daemon/certs/certs.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil {
+		return false, fmt.Errorf("Failed to read certificate file: %v", err)
+	}
+
+	return ensureCertsAreImported(systemStore, expectedCerts, existingCerts)
+}
+
+// readCertsFromFile reads in any PEM-encoded certificates from the given file
+func readCertsFromFile(path string) ([]*x509.Certificate, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open file: %v", err)
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			klog.Errorf("Failed to close file %s", f.Name())
+		}
+	}()
+
+	// Process the file one certificate at a time
+	scanner := bufio.NewScanner(f)
+	const endCertTag = "-----END CERTIFICATE-----"
+	scanner.Split(splitAt(endCertTag))
+	var certs []*x509.Certificate
+	for scanner.Scan() {
+		// Read next token
+		pemData := scanner.Text()
+		block, _ := pem.Decode([]byte(pemData))
+		if block == nil {
+			return nil, fmt.Errorf("No PEM data read from token")
+		}
+		// Decode and parse the certificate
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate: %v\n", err)
+		}
+		certs = append(certs, cert)
+	}
+	return certs, nil
+}
+
+// splitAt is a custom closure function to split data into complete PEM-encoded certificates
+func splitAt(substring string) func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	searchBytes := []byte(substring)
+	searchLen := len(searchBytes)
+	return func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		dataLen := len(data)
+		if atEOF && dataLen == 0 {
+			// Empty file, do nothing
+			return 0, nil, nil
+		}
+
+		// Find next separator and return the token
+		if i := bytes.Index(data, searchBytes); i >= 0 {
+			return i + searchLen, data[0 : i+searchLen], nil
+		}
+
+		if atEOF {
+			return dataLen, data, fmt.Errorf("Hit end of file without terminating the current token")
+		}
+		// Otherwise, continue reading file data
+		return 0, nil, nil
+	}
+}
+
+// getAllCerts reads all the certs from the given certificate store
+func getAllCerts(store windows.Handle) ([]*x509.Certificate, error) {
+	var existingSystemCerts []*x509.Certificate
+	var certContext *windows.CertContext
+	var err error
+	for {
+		certContext, err = windows.CertEnumCertificatesInStore(store, certContext)
+		if err != nil {
+			if errors.Is(err, windows.Errno(windows.CRYPT_E_NOT_FOUND)) {
+				// Error code implies we have read all certs:
+				// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certenumcertificatesinstore
+				break
+			}
+			return nil, fmt.Errorf("Error reading existing system certificate: %v", err)
+		}
+
+		// Convert the certificate bytes to Golang x509.Certificate
+		certBytes := unsafe.Slice(certContext.EncodedCert, certContext.Length)
+		x509Cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate from bytes %s: %v", certBytes, err)
```

Comment:
Is `certBytes` sensitive?

Reply:
+1

Reply:
potentially, I'll remove it from the error msg